librz/core/cbounds.c

```diff
@@ -0,0 +1,775 @@
+		if (!strstr(map->name, "heap") || !(map->perm & RZ_PERM_W)) {
+			continue;
+		}
+
+		boundaries.addr = map->addr;
+		boundaries.size = map->addr_end - map->addr;
+
+		if (!add_interval(core->io, list, boundaries, interval, map->perm)) {
+			rz_list_free(list);
+			return NULL;
+		}
+	}
+
+	return list;
+}
+
+/**
+ * \brief      Returns the RzDebug stack boundaries as a RzIOMap list
+ *
+ * \param      core        The RzCore to use
+ * \param      interval    The requested interval in RzIOMaps
+ *
+ * \return     On success a valid pointer (can be an empty list), otherwise NULL
+ */
+RZ_API RZ_OWN RzList /*<RzIOMap *>*/ *rz_core_get_boundaries_debug_stack(RZ_NONNULL RzCore *core, const RzInterval interval) {
+	rz_return_val_if_fail(core, NULL);
+
+	if (!core->bin->is_debugger) {
+		RZ_LOG_ERROR("core: no debugger connected for debug stack boundaries.\n");
+		return NULL;
+	}
+
+	// ensure data is syncronized
+	rz_debug_map_sync(core->dbg);
+
+	RzList *list = NULL;
+	RzListIter *iter = NULL;
+	RzDebugMap *map = NULL;
+	RzInterval boundaries;
+
+	// rz_io_map_free does not exist.
+	list = rz_list_newf(free);
+	if (!list) {
+		RZ_LOG_ERROR("core: failed to allocate RzList for debug stack boundaries.\n");
+		return NULL;
+	}
+
+	rz_list_foreach (core->dbg->maps, iter, map) {
+		if (!strstr(map->name, "stack")) {
+			continue;
+		}
+
+		boundaries.addr = map->addr;
+		boundaries.size = map->addr_end - map->addr;
+
+		if (!add_interval(core->io, list, boundaries, interval, map->perm)) {
+			rz_list_free(list);
+			return NULL;
+		}
+	}
+
+	return list;
+}
+
+/**
+ * \brief      Returns the RzDebug exec-set maps boundaries as a RzIOMap list
+ *
+ * \param      core        The RzCore to use
+ * \param      interval    The requested interval in RzIOMaps
+ *
+ * \return     On success a valid pointer (can be an empty list), otherwise NULL
+ */
+RZ_API RZ_OWN RzList /*<RzIOMap *>*/ *rz_core_get_boundaries_debug_program(RZ_NONNULL RzCore *core, const RzInterval interval) {
+	rz_return_val_if_fail(core, NULL);
+
+	if (!core->bin->is_debugger) {
+		RZ_LOG_ERROR("core: no debugger connected for debug program boundaries.\n");
+		return NULL;
+	}
+
+	// ensure data is syncronized
+	rz_debug_map_sync(core->dbg);
+
+	RzList *list = NULL;
+	RzListIter *iter = NULL;
+	RzDebugMap *map = NULL;
+	RzInterval boundaries;
+
+	// rz_io_map_free does not exist.
+	list = rz_list_newf(free);
+	if (!list) {
+		RZ_LOG_ERROR("core: failed to allocate RzList for debug program boundaries.\n");
+		return NULL;
+	}
+
+	rz_list_foreach (core->dbg->maps, iter, map) {
+		if (!(map->perm & RZ_PERM_X)) {
+			continue;
+		}
+
+		boundaries.addr = map->addr;
+		boundaries.size = map->addr_end - map->addr;
+
+		if (!add_interval(core->io, list, boundaries, interval, map->perm)) {
+			rz_list_free(list);
+			return NULL;
+		}
+	}
+
+	return list;
+}
+
+/**
+ * \brief      Returns a boundaries RzIOMap list based on the mode given (see [search|analysis|zoom].[in|from|to] options)
```

Comment:
I think description here should be more verbose in addition to `\brief`, since it's still unclear what these modes do exactly.

Reply:
yeah, it's cryptic for me too. i don't understand why we do this in the first place, but we can refactor this later.